### ISSUE TITLE
CA2222 false report for static methods (Do not decrease inherited member visibility)

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn-analyzers#1378]
#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Analyzer

DoNotDecreaseInheritedMemberVisibilityAnalyzer

#### Repro steps

```
public class A
{
    private static string GetType(string x)
    {
        return x.GetType().ToString();
    }
}
```

#### Expected behavior
No CA2222

#### Actual behavior
```
Severity	Code	Description	Project	File	Line	Suppression State
Warning	CA2222	Do not decrease inherited member visibility	ConsoleApp5	c:\users\mavasani\source\repos\ConsoleApp5\ConsoleApp5\Program.cs	3	Active
```

**Comment by sharwell** [comment]
In its current form, this seems like a duplicate of #1377.

❓ Why do we have CA2222? Is there any case where it will be reported but CS0114 (Member hides inherited member; missing override keyword) is not reported?

**Comment by mavasani** [comment]
It seems majority of cases where CA2222 is reported will have compiler error/warning, but the example listed at https://msdn.microsoft.com/en-us/library/ms182332.aspx has no compiler diagnostic. We should probably revisit whether we need this diagnostic at all.

**Comment by mavasani** [comment]
More notes:

1. Even VB seems to have compiler diagnostics for shadowing base members (we need to verify the overlap for VB diagnostic with CA2222)
2. Regardless of whether we keep or retire CA2222, the fixer should be retained.

**Comment by sharwell** [comment]
@mavasani In the example, the use of the `new` keyword is equivalent to an intentional suppression IMO. Unless we can find a case, I believe we should retire CA2222.

**Comment by mavasani** [comment]
@sharwell Another case where CA2222 is reported but there is no compiler warning:

1. Hiding method (private non-static): https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/DoNotDirectlyAwaitATaskTests.cs#L274
2. Inherited method (protected static): : https://github.com/dotnet/roslyn-analyzers/blob/master/src/Test.Utilities/DiagnosticAnalyzerTestBase.cs#L139

I think we should probably audit CA2222 for cases covered by existing compiler diagnostics and bail out for them, but still retain the diagnostic for remainder of cases.

**Comment by mavasani** [comment]
Marking this for discussion and tagging @jinujoseph @kuhlenh

**Comment by mavasani** [comment]
@sharwell Pointed out that our CA2222 implementation is buggy and should not fire for https://github.com/dotnet/roslyn-analyzers/issues/1378#issuecomment-345380212. I verified FXCop does not fire for this case and also verified that the [implementation](http://index/?query=DoNotDecreaseInheritedMemberVisibility&rightProject=UsageRules&file=DoNotDecreaseInheritedMemberVisibility.cs&line=31) checks for identical signatures. This makes a good case to deprecate CA2222 in favor of compiler diagnostics, and just retain the fixer and add FixAll support.

**Comment by mavasani** [comment]
Tagging @nguerrera if he recalls any specific reason/scenario for having [CA2222](https://msdn.microsoft.com/en-us/library/ms182332.aspx) in FXCop other than the fact that FXCop was MSIL based and did not have duplication with compiler diagnostics.

**Comment by 333fred** [comment]
I talked with @nguerrera earlier. For this rule, the idea is that even with the new keyword, you aren't actually hiding anything. The base method will still be visible to consumers of the class, even if you use the new keyword. I'm not sure that, given the presence of the new keyword, that's enough reason to keep CA2222 around. @mavasani and @sharwell, thoughts?

**Comment by mavasani** [comment]
I agree with you and @sharwell. CA2222 doesn't seem to provide too much of a value add over compiler diagnostics, and we can get it back in future if we get feedback.

NOTE: We would need to change the fixer to work on compiler diagnostics.

**Comment by 333fred** [comment]
@mavasani there already appears to be an ide fixer for adding the new keyword. Should we be continuing to fire about increasing accessibility, since CS0108 does not? As an example:
```
public class Base
{
	public void M1() {}
}
public class Derived : Base
{
	// IDE fixer for adding new, which suppresses the violation, and our fixer for changing to public, which does not suppress the violation
	private void M1() {}
}
```
```
public class Base
{
	public void M1() {}
}
public class Derived : Base
{
	// We would have fired CA2222 here and offered to increase visibility, no warning now
	private new void M1() {}
}
```

Namely, the fix for CA2222 doesn't seem too relevant to the original.

**Comment by mavasani** [comment]
> there already appears to be an ide fixer for adding the new keyword

In that case, we can remove our fixer as well.
